### PR TITLE
Eng 13840 debug build failed assertion for select subquery

### PR DESCRIPTION
**Pull request by kaikaizi**
Problem fixed with Calcite planner. (That is, if you turn the secret switch on to use Calcite planner and execute the Calcite plan, the problematic query works). The legacy planner still need to be fixed, before it can be used in production code.

Calcite originally works on this type of query; but generates an additional SINGLE_VALUE aggregation node that doesn't do calculations; and VoltDB does not have a corresponding aggregation node. This fix simply removes that aggregation node that sits on top of a Calc node in logical phase.

### CODE REVIEW

File: src/frontend/org/voltdb/expressions/AggregateExpression.java [hunk 1]
```diff
@@ -53,12 +53,11 @@ public int getUserAggregateId() {
 
     @Override
     public boolean equals(Object obj) {
```
Comment (kaikaizi) [inline src/frontend/org/voltdb/expressions/AggregateExpression.java]: All changes in this file are stylistic changes only.

File: src/frontend/org/voltdb/plannerv2/rules/PlannerRules.java [hunk 2]
```diff
@@ -166,6 +141,7 @@ public RuleSet getRules() {
             VoltLTableScanRule.INSTANCE,
             VoltLCalcRule.INSTANCE,
             VoltLAggregateRule.INSTANCE,
+            VoltLAggregateCalcMergeRule.INSTANCE,       // eliminate Calcite's SINGLE_VALUE aggregation
```
Comment (alexeevm) [inline src/frontend/org/voltdb/plannerv2/rules/PlannerRules.java]: Should this rule be a part of the HEP_OUTER_JOIN phase? We want it to succeed no matter what, right?

Comment (kaikaizi) [reply src/frontend/org/voltdb/plannerv2/rules/PlannerRules.java]: Although the example query is a NLJ with sub queries, I think SINGLE_VALUE aggregation does not necessarily need a join node to be applicable.
Putting it in the `LOGICAL` ruleset still makes sure that the rule is applied for all queries, just like `HEP_OUTER_JOIN` phase, is that right?

Comment (alexeevm) [reply src/frontend/org/voltdb/plannerv2/rules/PlannerRules.java]: Yes, this is correct. The only difference between the HEP and Volcano rules / planners is that the latter uses cost to find the best rule - the rule may be applied but it does not guarantee that the transformed node will be a winner. The HEP simply applies it regardless of the cost. If the rule is applied, the outcome guaranteed to have a transformed node

Comment (kaikaizi) [reply src/frontend/org/voltdb/plannerv2/rules/PlannerRules.java]: I see. The key is in `VoltPlanner.transformHep(...)` stage that forces it to be used irrespective of cost.
Thanks, will make the change in my next PR.

Comment (kaikaizi) [reply src/frontend/org/voltdb/plannerv2/rules/PlannerRules.java]: Update:
From what I see, I cannot get it to work with a Hep rule.
If I move `VoltLAggregateCalcMergeRule.INSTANCE` to the last of `HEP_OUTER_JOIN` set, although I can see the rule get registered, it is never used;
If I add a new `PlannerRule#Phase` that returns a list of a single `VoltLAggregateCalcMergeRule.INSTANCE`, I get an exception from Calcite:

> Node [rel#512:Subset#36.CONVENTION.[].any] could not be implemented; planner state:
> Root: rel#512:Subset#36.CONVENTION.[].any
> Original rel:
> VoltLogicalCalc(subset=[rel#512:Subset#36.CONVENTION.[].any], expr#0..1=[{inputs}], I=[$t0]): rowcount = 1.875E23, cumulative cost = {1.875E23 rows, 1.875E23 cpu, 0.0 io}, id = 510
>   VoltLogicalJoin(subset=[rel#509:Subset#35.CONVENTION.[].single], condition=[true], joinType=[inner]): rowcount = 1.875E23, cumulative cost = {1.875E23 rows, 1.875E23 cpu, 0.0 io}, id = 508
>     VoltLogicalTableScan(subset=[rel#481:Subset#21.CONVENTION.[].single], table=[[public, R2]]): rowcount = 1000000.0, cumulative cost = {100.0 rows, 100.0 cpu, 0.0 io}, id = 323
>     VoltLogicalAggregate(subset=[rel#507:Subset#34.CONVENTION.[].single], group=[{}], EXPR$0=[MAX($0)]): rowcount = 1.875E17, cumulative cost = {2.109375E17 rows, 2.109375E17 cpu, 0.0 io}, id = 506
>       VoltLogicalCalc(subset=[rel#505:Subset#33.CONVENTION.[].single], expr#0..2=[{inputs}], I=[$t0]): rowcount = 1.875E18, cumulative cost = {1.875E18 rows, 1.875E18 cpu, 0.0 io}, id = 504
>         VoltLogicalJoin(subset=[rel#503:Subset#32.CONVENTION.[].single], condition=[AND(=($0, $1), >($0, $2))], joinType=[inner]): rowcount = 1.875E18, cumulative cost = {1.875E18 rows, 1.875E18 cpu, 0.0 io}, id = 502
>           VoltLogicalTableScan(subset=[rel#481:Subset#21.CONVENTION.[].single], table=[[public, R2]]): rowcount = 1000000.0, cumulative cost = {100.0 rows, 100.0 cpu, 0.0 io}, id = 323
>           VoltLogicalAggregate(subset=[rel#501:Subset#31.CONVENTION.[].single], group=[{0}], agg#0=[SINGLE_VALUE($1)]): rowcount = 2.5E13, cumulative cost = {2.8125E13 rows, 2.8125E13 cpu, 0.0 io}, id = 500
>             VoltLogicalCalc(subset=[rel#499:Subset#30.CONVENTION.[].single], expr#0..2=[{inputs}], I1=[$t2], EXPR$0=[$t0]): rowcount = 2.5E14, cumulative cost = {2.5E14 rows, 2.5E14 cpu, 0.0 io}, id = 498
>               VoltLogicalSort(subset=[rel#497:Subset#29.CONVENTION.[1].single], sort0=[$1], dir0=[ASC]): rowcount = 2.5E14, cumulative cost = {2.5E14 rows, 2.5E14 cpu, 0.0 io}, id = 496
>                 VoltLogicalCalc(subset=[rel#495:Subset#28.CONVENTION.[].single], expr#0..2=[{inputs}], EXPR$0=[$t2], I=[$t0], I1=[$t1]): rowcount = 2.5E14, cumulative cost = {2.5E14 rows, 2.5E14 cpu, 0.0 io}, id = 494
>                   VoltLogicalAggregate(subset=[rel#493:Subset#27.CONVENTION.[].single], group=[{0, 1}], EXPR$0=[COUNT($2)]): rowcount = 2.5E14, cumulative cost = {2.8125E14 rows, 2.8125E14 cpu, 0.0 io}, id = 492
>                     VoltLogicalCalc(subset=[rel#491:Subset#26.CONVENTION.[].single], expr#0..2=[{inputs}], I=[$t0], I1=[$t2], $f1=[$t2]): rowcount = 2.5E15, cumulative cost = {2.5E15 rows, 2.5E15 cpu, 0.0 io}, id = 490
>                       VoltLogicalJoin(subset=[rel#489:Subset#25.CONVENTION.[].single], condition=[true], joinType=[inner]): rowcount = 2.5E15, cumulative cost = {2.5E15 rows, 2.5E15 cpu, 0.0 io}, id = 488
>                         VoltLogicalJoin(subset=[rel#486:Subset#24.CONVENTION.[].single], condition=[CASE(IS NULL($0), IS NULL($1), IS NULL($1), IS NULL($0), =(CAST($0):INTEGER NOT NULL, CAST($1):INTEGER NOT NULL))], joinType=[inner]): rowcount = 2.5E10, cumulative cost = {2.5E10 rows, 2.5E10 cpu, 0.0 io}, id = 485
>                           VoltLogicalTableScan(subset=[rel#482:Subset#22.CONVENTION.[].single], table=[[public, R1]]): rowcount = 1000000.0, cumulative cost = {100.0 rows, 100.0 cpu, 0.0 io}, id = 338
>                           VoltLogicalAggregate(subset=[rel#484:Subset#23.CONVENTION.[].single], group=[{0}]): rowcount = 100000.0, cumulative cost = {100000.0 rows, 100000.0 cpu, 0.0 io}, id = 483
>                             VoltLogicalTableScan(subset=[rel#481:Subset#21.CONVENTION.[].single], table=[[public, R2]]): rowcount = 1000000.0, cumulative cost = {100.0 rows, 100.0 cpu, 0.0 io}, id = 323
>                         VoltLogicalAggregate(subset=[rel#484:Subset#23.CONVENTION.[].single], group=[{0}]): rowcount = 100000.0, cumulative cost = {100000.0 rows, 100000.0 cpu, 0.0 io}, id = 483
>                           VoltLogicalTableScan(subset=[rel#481:Subset#21.CONVENTION.[].single], table=[[public, R2]]): rowcount = 1000000.0, cumulative cost = {100.0 rows, 100.0 cpu, 0.0 io}, id = 323
> 
> Sets:
> Set#21, type: RecordType(INTEGER I)
> 	rel#481:Subset#21.CONVENTION.[].single, best=rel#323, importance=0.0
> 		rel#323:VoltLogicalTableScan.CONVENTION.[].single(table=[public, R2]), rowcount=1000000.0, cumulative cost={100.0 rows, 100.0 cpu, 0.0 io}
> Set#22, type: RecordType(INTEGER I)
> 	rel#482:Subset#22.CONVENTION.[].single, best=rel#338, importance=0.0
> 		rel#338:VoltLogicalTableScan.CONVENTION.[].single(table=[public, R1]), rowcount=1000000.0, cumulative cost={100.0 rows, 100.0 cpu, 0.0 io}
> Set#23, type: RecordType(INTEGER I)
> 	rel#484:Subset#23.CONVENTION.[].single, best=rel#483, importance=0.0
> 		rel#483:VoltLogicalAggregate.CONVENTION.[].single(input=rel#481:Subset#21.CONVENTION.[].single,group={0}), rowcount=100000.0, cumulative cost={100100.0 rows, 100100.0 cpu, 0.0 io}
> Set#24, type: RecordType(INTEGER I, INTEGER I0)
> 	rel#486:Subset#24.CONVENTION.[].single, best=rel#485, importance=0.0
> 		rel#485:VoltLogicalJoin.CONVENTION.[].single(left=rel#482:Subset#22.CONVENTION.[].single,right=rel#484:Subset#23.CONVENTION.[].single,condition=CASE(IS NULL($0), IS NULL($1), IS NULL($1), IS NULL($0), =(CAST($0):INTEGER NOT NULL, CAST($1):INTEGER NOT NULL)),joinType=inner), rowcount=2.5E10, cumulative cost={2.50001002E10 rows, 2.50001002E10 cpu, 0.0 io}
> Set#25, type: RecordType(INTEGER I, INTEGER I0, INTEGER I1)
> 	rel#489:Subset#25.CONVENTION.[].single, best=rel#488, importance=0.0
> 		rel#488:VoltLogicalJoin.CONVENTION.[].single(left=rel#486:Subset#24.CONVENTION.[].single,right=rel#484:Subset#23.CONVENTION.[].single,condition=true,joinType=inner), rowcount=2.5E15, cumulative cost={2.5000250002003E15 rows, 2.5000250002003E15 cpu, 0.0 io}
> Set#26, type: RecordType(INTEGER I, INTEGER I1, INTEGER $f1)
> 	rel#491:Subset#26.CONVENTION.[].single, best=rel#490, importance=0.0
> 		rel#490:VoltLogicalCalc.CONVENTION.[].single(input=rel#489:Subset#25.CONVENTION.[].single,expr#0..2={inputs},I=$t0,I1=$t2,$f1=$t2), rowcount=2.5E15, cumulative cost={5.0000250002003E15 rows, 5.0000250002003E15 cpu, 0.0 io}
> Set#27, type: RecordType(INTEGER I, INTEGER I1, BIGINT EXPR$0)
> 	rel#493:Subset#27.CONVENTION.[].single, best=rel#492, importance=0.0
> 		rel#492:VoltLogicalAggregate.CONVENTION.[].single(input=rel#491:Subset#26.CONVENTION.[].single,group={0, 1},EXPR$0=COUNT($2)), rowcount=2.5E14, cumulative cost={5.2812750002003E15 rows, 5.2812750002003E15 cpu, 0.0 io}
> Set#28, type: RecordType(BIGINT EXPR$0, INTEGER I, INTEGER I1)
> 	rel#495:Subset#28.CONVENTION.[].single, best=rel#494, importance=0.0
> 		rel#494:VoltLogicalCalc.CONVENTION.[].single(input=rel#493:Subset#27.CONVENTION.[].single,expr#0..2={inputs},EXPR$0=$t2,I=$t0,I1=$t1), rowcount=2.5E14, cumulative cost={5.5312750002003E15 rows, 5.5312750002003E15 cpu, 0.0 io}
> Set#29, type: RecordType(BIGINT EXPR$0, INTEGER I, INTEGER I1)
> 	rel#497:Subset#29.CONVENTION.[1].single, best=rel#496, importance=0.0
> 		rel#496:VoltLogicalSort.CONVENTION.[1].single(input=rel#495:Subset#28.CONVENTION.[].single,sort0=$1,dir0=ASC), rowcount=2.5E14, cumulative cost={5.7812750002003E15 rows, 5.7812750002003E15 cpu, 0.0 io}
> Set#30, type: RecordType(INTEGER I1, BIGINT EXPR$0)
> 	rel#499:Subset#30.CONVENTION.[].single, best=rel#498, importance=0.0
> 		rel#498:VoltLogicalCalc.CONVENTION.[].single(input=rel#497:Subset#29.CONVENTION.[1].single,expr#0..2={inputs},I1=$t2,EXPR$0=$t0), rowcount=2.5E14, cumulative cost={6.0312750002003E15 rows, 6.0312750002003E15 cpu, 0.0 io}
> Set#31, type: RecordType(INTEGER I1, BIGINT $f1)
> 	rel#501:Subset#31.CONVENTION.[].single, best=rel#500, importance=0.0
> 		rel#500:VoltLogicalAggregate.CONVENTION.[].single(input=rel#499:Subset#30.CONVENTION.[].single,group={0},agg#0=SINGLE_VALUE($1)), rowcount=2.5E13, cumulative cost={6.0594000002003E15 rows, 6.0594000002003E15 cpu, 0.0 io}
> Set#32, type: RecordType(INTEGER I, INTEGER I1, BIGINT $f1)
> 	rel#503:Subset#32.CONVENTION.[].single, best=rel#502, importance=0.0
> 		rel#502:VoltLogicalJoin.CONVENTION.[].single(left=rel#481:Subset#21.CONVENTION.[].single,right=rel#501:Subset#31.CONVENTION.[].single,condition=AND(=($0, $1), >($0, $2)),joinType=inner), rowcount=1.875E18, cumulative cost={1.88105940000020019E18 rows, 1.88105940000020019E18 cpu, 0.0 io}
> Set#33, type: RecordType(INTEGER I)
> 	rel#505:Subset#33.CONVENTION.[].single, best=rel#504, importance=0.0
> 		rel#504:VoltLogicalCalc.CONVENTION.[].single(input=rel#503:Subset#32.CONVENTION.[].single,expr#0..2={inputs},I=$t0), rowcount=1.875E18, cumulative cost={3.7560594000002002E18 rows, 3.7560594000002002E18 cpu, 0.0 io}
> Set#34, type: RecordType(INTEGER EXPR$0)
> 	rel#507:Subset#34.CONVENTION.[].single, best=rel#506, importance=0.0
> 		rel#506:VoltLogicalAggregate.CONVENTION.[].single(input=rel#505:Subset#33.CONVENTION.[].single,group={},EXPR$0=MAX($0)), rowcount=1.875E17, cumulative cost={3.9669969000002002E18 rows, 3.9669969000002002E18 cpu, 0.0 io}
> Set#35, type: RecordType(INTEGER I, INTEGER EXPR$0)
> 	rel#509:Subset#35.CONVENTION.[].single, best=rel#508, importance=0.0
> 		rel#508:VoltLogicalJoin.CONVENTION.[].single(left=rel#481:Subset#21.CONVENTION.[].single,right=rel#507:Subset#34.CONVENTION.[].single,condition=true,joinType=inner), rowcount=1.875E23, cumulative cost={1.875039669969E23 rows, 1.875039669969E23 cpu, 0.0 io}
> Set#36, type: RecordType(INTEGER I)
> 	rel#511:Subset#36.CONVENTION.[].single, best=rel#510, importance=0.0
> 		rel#510:VoltLogicalCalc.CONVENTION.[].single(input=rel#509:Subset#35.CONVENTION.[].single,expr#0..1={inputs},I=$t0), rowcount=1.875E23, cumulative cost={3.7500396699690003E23 rows, 3.7500396699690003E23 cpu, 0.0 io}
> 	rel#512:Subset#36.CONVENTION.[].any, best=null, importance=0.0
Any idea why is that the case?

Comment (alexeevm) [reply src/frontend/org/voltdb/plannerv2/rules/PlannerRules.java]: `super(operand(LogicalAggregate.class, operand(LogicalCalc.class, none())));`
After the logical phase it should be VoltLogicalAggregate and VoltLogicalCalc. VoltLogicalAggregate extends Aggregate but not the LogicalAggregate. The same goes for the VoltLogicalCalc which extends Calc and not the LogicalCalc

Comment (kaikaizi) [reply src/frontend/org/voltdb/plannerv2/rules/PlannerRules.java]: Thanks for the explanation. I remember running into this at one point (probably when  working on MP rule).
The changes are in the last commit of #6665. (The #6665 itself might also be interesting, as it turns on legacy planner's plan node validation as part of planning).

**Pull request merged**